### PR TITLE
Reposition cursors after formatting

### PR DESCRIPTION
**Pull request by faergeek**
I think I figured out an optimal way for repositioning cursors after formatting. I didn't test it in everyday work yet, but I'm going to start doing that now. Can you please take a look and tell what you think about it?

By the way, did you have a chance to try out `partial_update = "diff"` and see if it works okay? I personally used it day-to-day after I first implemented it and since the last fix was merged I didn't encounter any problems.

### CODE REVIEW

**Comment by faergeek** [comment]
Just changed the implementation to be really straight forward and less prone to errors. Now I'm not sure if the tests I added are really needed. And the one I added before does not really test what it says it tests, because added `interhunkctxlen` options squashes together hunks which are less lines apart than the provided value.

**Comment by elentok** [comment]
Hey Sergey :)

Sorry for the delayed response, I was swamped with work.

I've been using `partial_update = "diff"` since I merged your PR and it's been working great 👍 
After we disabled cursor restoration I did notice that when using splits the other windows hosting the same buffer would jump to the top so this PR is very very welcome 🙏 

If I understand correctly you're only restoring cursors that were instead a deleted hunk, right? is that what's causing the other windows to jump to the top?

Anyway, looks great, do you think going to a char-level diff is worth the trouble (and possible performance penatly)? what happens if when a hunk is deleted you just jump to the beginning of where it was or something similar?

Another thing, what if we use vim.fn.setline instead of vim.api.nvim_buf_set_lines? from what I understand it doesn't reset jumps and extmarks and the line (I think).

Thanks a lot for this! I really appreciate it and I can't wait to try it out 👍

**Comment by faergeek** [comment]
>After we disabled cursor restoration I did notice that when using splits the other windows hosting the same buffer would jump to the top so this PR is very very welcome 🙏

Sadly, this PR does not seem to solve this exact problem. Also I researched a bit and couldn't find an api to fix window position properly. I guess I need to take a closer look at the autocommand I disabled, may be I'll find a solution there :-)

>If I understand correctly you're only restoring cursors that were instead a deleted hunk, right? is that what's causing the other windows to jump to the top?

In this PR I adjust cursor positions in windows hosting the buffer. But in some cases windows jump to the top anyway, until they receive focus. Only after receiving focus windows scroll to the proper position to show the cursor. I found ways to adjust window position as well, but only for current window. And I didn't want to focus every window just to do that. But that seems to be the only way to fix this problem.

>do you think going to a char-level diff is worth the trouble (and possible performance penatly)?

Not really sure about performance penalty with the current approach. Before this solution I tried a dynamic programming algorithm, which is linked in Emacs plugin's readme (Needleman-Wunsch algorithm for sequence alignment). It has a quadratic complexity in time and space and I tested it on a file with ~1500 lines of code, where I just indented all of the lines once and then formatted it back with prettier. That one had a horrible performance for sure) On second try neovim was killed because it used too much memory)

But the current solution seems to be really fast. Although I encountered some edge cases, where cursor is not repositioned correctly because of the way hunks are calculated. For that reason now I'm figuring out Hirschberg algorithm, which is based on Needleman-Wunsch, but is quadratic only in time and linear in space and it should perform much better. Plus earlier I applied an algorithm on the whole hunks on line level and now I plan to only apply it to hunks on char level and only those where replacement happens, not add-only or delete-only. And I want to only apply it if the ranges are really small, like 10, may be 20 characters. I hope it would not perform too bad.

But that said, if performance with the current solution is already not great (which I didn't notice yet, so please confirm), then there's definitely no reason to try a more complex one :-)

>what happens if when a hunk is deleted you just jump to the beginning of where it was or something similar?

Not sure what you mean. In case of line hunks, if there are only additions or only deletions, we don't touch cursors at all and rely on the default behavior. All of that added logic runs only for "sub(stitute)" hunks (addition + replacement).

>Another thing, what if we use vim.fn.setline instead of vim.api.nvim_buf_set_lines? from what I understand it doesn't reset jumps and extmarks and the line (I think).

Not sure about that. I only tried `nvim_buf_set_text`, which doesn't replace whole lines, but a range from character to character. I hoped it would reposition cursors and extmarks by itself, but sadly it breaks at least in case where newline right before cursor or extmark is deleted. May be `vim.fn.setline` doesn't suffer from such problems, but I would not bet on that to be honest)

>Thanks a lot for this! I really appreciate it and I can't wait to try it out 👍

Thanks :+1: 

>EDIT: Forgot to mention, the code itself looks great, I just left a small comment about splitting up the function.

That's strange, I don't see any comments on code. May be it's because the PR is in a draft state?

**Review by elentok — COMMENTED** [review]
Oops, forgot to submit my review 🤦

File: lua/format-on-save/update-buffer.lua [hunk 1]
```diff
@@ -1,39 +1,167 @@
...
+    del_count = dc,
+    del_end = ds + dc - 1,
+  }
+
+  return result
+end
+
 ---@param original_lines string[]
 ---@param formatted_lines string[]
 local function update_buffer_with_diff(original_lines, formatted_lines)
```
[the remark this answers is not in the document]
Comment (faergeek) [reply lua/format-on-save/update-buffer.lua:70]: The ordering is a bit different, but I got the idea. I'll update it a bit later

**Comment by elentok** [comment]
> > After we disabled cursor restoration I did notice that when using splits the other windows hosting the same buffer would jump to the top so this PR is very very welcome 🙏
> 
> Sadly, this PR does not seem to solve this exact problem. Also I researched a bit and couldn't find an api to fix window position properly. I guess I need to take a closer look at the autocommand I disabled, may be I'll find a solution there :-)
> 

👍 

> > If I understand correctly you're only restoring cursors that were instead a deleted hunk, right? is that what's causing the other windows to jump to the top?
> 
> In this PR I adjust cursor positions in windows hosting the buffer. But in some cases windows jump to the top anyway, until they receive focus. Only after receiving focus windows scroll to the proper position to show the cursor. I found ways to adjust window position as well, but only for current window. And I didn't want to focus every window just to do that. But that seems to be the only way to fix this problem.
> 
> > do you think going to a char-level diff is worth the trouble (and possible performance penatly)?
> 
> Not really sure about performance penalty with the current approach. Before this solution I tried a dynamic programming algorithm, which is linked in Emacs plugin's readme (Needleman-Wunsch algorithm for sequence alignment). It has a quadratic complexity in time and space and I tested it on a file with ~1500 lines of code, where I just indented all of the lines once and then formatted it back with prettier. That one had a horrible performance for sure) On second try neovim was killed because it used too much memory)
> 
> But the current solution seems to be really fast. Although I encountered some edge cases, where cursor is not repositioned correctly because of the way hunks are calculated. For that reason now I'm figuring out Hirschberg algorithm, which is based on Needleman-Wunsch, but is quadratic only in time and linear in space and it should perform much better. Plus earlier I applied an algorithm on the whole hunks on line level and now I plan to only apply it to hunks on char level and only those where replacement happens, not add-only or delete-only. And I want to only apply it if the ranges are really small, like 10, may be 20 characters. I hope it would not perform too bad.
> 
> But that said, if performance with the current solution is already not great (which I didn't notice yet, so please confirm), then there's definitely no reason to try a more complex one :-)
> 

I haven't noticed any performance issues so far, was just wondering 👍 

I'm still not 100% clear as to why the character-wise diff is needed, isn't it enough to compare lines, replace the hunks, if the cursor was inside a replaced hunk just move it to the beginning of the new hunk and be done with it?

I'm willing to merge this but I'm worried about the complexity and future maintenance... what do you think?

> > what happens if when a hunk is deleted you just jump to the beginning of where it was or something similar?
> 
> Not sure what you mean. In case of line hunks, if there are only additions or only deletions, we don't touch cursors at all and rely on the default behavior. All of that added logic runs only for "sub(stitute)" hunks (addition + replacement).
> 
> > Another thing, what if we use vim.fn.setline instead of vim.api.nvim_buf_set_lines? from what I understand it doesn't reset jumps and extmarks and the line (I think).
> 
> Not sure about that. I only tried `nvim_buf_set_text`, which doesn't replace whole lines, but a range from character to character. I hoped it would reposition cursors and extmarks by itself, but sadly it breaks at least in case where newline right before cursor or extmark is deleted. May be `vim.fn.setline` doesn't suffer from such problems, but I would not bet on that to be honest)
> 
> > Thanks a lot for this! I really appreciate it and I can't wait to try it out 👍
> 
> Thanks 👍
> 
> > EDIT: Forgot to mention, the code itself looks great, I just left a small comment about splitting up the function.
> 
> That's strange, I don't see any comments on code. May be it's because the PR is in a draft state?

I forgot to submit my review comments 🤦

**Comment by faergeek** [comment]
>I'm still not 100% clear as to why the character-wise diff is needed, isn't it enough to compare lines, replace the hunks, if the cursor was inside a replaced hunk just move it to the beginning of the new hunk and be done with it?

Imagine this case. You are in the process of changing some lua code that looks like this:

```lua
if something_long == "long" then
end
```

You need to add a few more conditions to the `if` and you're not yet sure how long those conditions are, so you just start adding them at the same line, before `then`. You add just one of them and realize that the line became too long and you'd like to format it first, before adding more conditions. So you stopped with the code that looks like this:

```lua
if something_long == "long" and not (something_else == "this is really really long") then
                                                                                -- ^ cursor is here
end
```

And then you simply run it through the formatter, which turns it into this:

```lua
  if
-- ^ cursor ends up being here (without this PR)
-- or in case of the solution you suggest it would end up
-- right at the start of the line with "if" because
-- that's where the beginning of the new hunk is
    something_long == 'long'
    and not (something_else == 'this is really really long')
                                                        -- ^ but it would be much nicer
                                                        -- if it would end up right here,
                                                        -- where it logically was before,
                                                        -- so you could just continue to add
                                                        -- all the conditions you wanted
  then
  end
```

So the cursor ends up in a different place after formatting, because one line was replaced with four lines.

I often find myself in cases like this. Sometimes I like to just spit out some code in one go and I'm not sure how long it would be before I start. And in the middle of writing I realize that the line is now too long and it's inconvenient to continue that way, so I want to format it first, but without then needing to manually move cursor to the position I was in. I hope that makes sense. You can also think about some fluent apis or something like that, where it's very likely to happen when you start on a single line, add some more calls and you need to break that into lines first and then add more.

These are the kinds of cases I would like to address with this pull request.

And so to move the cursor to where it's supposed to go, we need to know how the characters were changed by the formatter. With that information we can then understand how to move cursor forward or backward.

>I'm willing to merge this but I'm worried about the complexity and future maintenance... what do you think?

Just in case, this is not meant to be merged yet :-) And about complexity. I'm actually exploring a solution, which might be a little bit more complicated and it should be added it on top of the one in this PR, for cases where more precision is needed.

And I'm not sure how to go about it if the current complexity is already too much. I can only break it down into more manageable pieces, which would allow us to test them separately. But complexity would not go anywhere, it would just be abstracted a bit, so that on the high level it stays simple.

I could provide both the example of a problem and the solution, if you're interested. Although an exact example I have right now (for which I have a test locally) might seem not to be worth the effort :-)

**Comment by elentok** [comment]
Wow! that's quite impressive! I don't have an objection to merging, it sounds very helpful.

Perhaps a possible solution might be to make the char-wise diff optional? so users can choose if they want it or not?

perhaps the `update_buffer` config flag should become something more complex? e.g.

```lua
format_on_save.setup({
  update_buffer: {
    diff = true,
    cursor_restore = "charwise", -- or "linewise" or false
  },
})
```

Or maybe switch to a set of general experiment flags:

```lua
format_on_save.setup({
  experimental: {
    update_buffer_with_diff = true,
    restore_charwise_cursor = true,
  }
})
```

Not sure, just brainstorming. The idea behind it all is to be able to test out experimental stuff without breaking the plugin and being able to go back to a stable version by disabling experiments (instead of by locking to a specific version).

What do you think?

**Comment by faergeek** [comment]
>Wow! that's quite impressive! I don't have an objection to merging, it sounds very helpful.

Glad to hear that :+1: 

About the flags, I'm all for it, we definitely need to test the performance of the solution for repositioning cursors. I think this one is the way to go:

```lua
format_on_save.setup({
  experimental: {
    update_buffer_with_diff = true,
    restore_charwise_cursor = true,
  }
})
```

The only confusing thing is that `restore_charwise_cursor` would depend on `update_buffer_with_diff` to be enabled.

For me it seems like we should try to make sure that "diff" mode can be left as the only mode first and only then get back to this PR. I believe the only problem is related to the "window jumping" you mentioned. If there's anything else, please mention it as well.

So, what would I suggest to do next:

* Figure out the solution to avoid other windows jumping to the top, most probably based on the current one with `partial_update = true`.
* Use it for some more time to make sure there's no more issues.
* When all is confirmed to work great, make "diff" the only way to apply updates, remove the other two.

And then (or in the meantime) get back to this PR and:

* Most probably rebase it, clean up the code a bit, break it down into more manageable parts
* Add tests, which would justify the need for Hirschberg algorithm and then apply the algorithm to make them pass
* Merge it after a bit of testing, behind an experimental flag

And then after some time we'll probably remove the flag and just enable it by default as well.

What do you think?

**Comment by elentok** [comment]
Sounds great, I'll handle the flags refactor this weekend 👍

**Comment by faergeek** [comment]
@elentok Did you notice anything special about cases where this jumping happens? I'm trying to reproduce it now, but I can't. May be formatting should happen at a particular point relative to split/cursor position or something like that? Anything would help.

>After we disabled cursor restoration I did notice that when using splits the other windows hosting the same buffer would jump to the top so this PR is very very welcome

**Comment by elentok** [comment]
Hmm, I tried to reproduce again and couldn't, I'm not sure what happened exactly, I'll keep an eye on it.

I made a couple of commits that add the new experiment system, I split the "partial_update" into "partial_update" and "disable_restore_cursors" flags to allow playing with both settings separately.

I also found a way to get the restore cursors to preserve the scroll by wrapping it with a call to `vim.fn.timerstart(0, ...)` (the 
old javascript setTimeout(function(), 0) trick 😄 ).

**Comment by faergeek** [comment]
> I made a couple of commits that add the new experiment system, I split the "partial_update" into "partial_update" and "disable_restore_cursors" flags to allow playing with both settings separately.
> 
> I also found a way to get the restore cursors to preserve the scroll by wrapping it with a call to `vim.fn.timerstart(0, ...)` (the old javascript setTimeout(function(), 0) trick 😄 ).

Yeah, there's no async problem in JS that cannot be solved with setTimeout(..., 0), except the problem of too many setTimeout(..., 0) 😄

By the way, when I looked at that cursor restoration, I noticed that you save cursors in `format` call, but restore them in a separate call in `autocmd`. But in this PR I move cursors right at the end of formatting, without needing autocmd to fire. Is there a reason to do it in a separate `autocmd` instead? In that case it does not do anything with cursors if you simply invoke `Format` without really saving. I understand that the plugin is concentrated on formatting on save, but anyway this seems to be a downside. Is there also an upside?

**Comment by elentok** [comment]
> > I made a couple of commits that add the new experiment system, I split the "partial_update" into "partial_update" and "disable_restore_cursors" flags to allow playing with both settings separately.
> > I also found a way to get the restore cursors to preserve the scroll by wrapping it with a call to `vim.fn.timerstart(0, ...)` (the old javascript setTimeout(function(), 0) trick 😄 ).
> 
> Yeah, there's no async problem in JS that cannot be solved with setTimeout(..., 0), except the problem of too many setTimeout(..., 0) 😄
> 
> By the way, when I looked at that cursor restoration, I noticed that you save cursors in `format` call, but restore them in a separate call in `autocmd`. But in this PR I move cursors right at the end of formatting, without needing autocmd to fire. Is there a reason to do it in a separate `autocmd` instead? In that case it does not do anything with cursors if you simply invoke `Format` without really saving. I understand that the plugin is concentrated on formatting on save, but anyway this seems to be a downside. Is there also an upside?

That's a good question, I don't remember why I used the post-write autocmd, I'll try moving it to the post-format stage and see what happens.

**Comment by elentok** [comment]
I figured out the cursor restoration issue I was noticing, it's actually only broken on LSP formatters 😄 

Your diff mode changes actually work better than the LSP 👍 

Maybe I should add the ability to only restore the cursor when using LSP, wdyt?

**Comment by faergeek** [comment]
> Maybe I should add the ability to only restore the cursor when using LSP, wdyt?

I'm not really sure about that. I heard that some language servers effectively just do whole buffer replacements, similar to how shell formatters work, which is bad, because we can't apply similar diffing to them without duplicating pretty much all of the logic from `vim.lsp.buf.format`. But others do more fine grained edits. And it means that there better should be an option per lsp formatter, not the global one. So that cursor restoration can be enabled only for those formatters that don't do granular edits. And then the user should determine for which formatters it should be enabled, which is also not very cool.

May be after all it's worth looking into how `vim.lsp.buf.format` is implemented and see how bad it would be to reimplement it on our side with diffing. At least there exists a `vim.lsp.util.make_formatting_params` helper, may be that would help somewhat)

**Comment by faergeek** [comment]
By the way, with the approach currently implemented for adjusting the cursor position, if you undo the changes made by the formatter and then redo them, these adjustments are gone and cursor goes where it would go without those adjustments. I'm not sure if it matters much in practice, but I suppose the only way to fix that would be to calculate character level diffs not only for hunks where any of the cursors happen to be. If only `nvim_buf_set_text` would always reposition properly, I guess it then there would be no problems with losing marks, extmarks, etc. But in that case potential performance penalty could be much higher. At least in worst cases when there are a lot of changes by formatter at once. Though I'm pretty sure that such cases should be considered exceptional, because usually you don't reformat the whole file constantly and most changes would be gone on the line level already.

**Comment by elentok** [comment]
Hi :)

Amazing work!!! haven't had much time today, I'll go over it more in depth tomorrow but it looks great!

Thanks!!

**Comment by faergeek** [comment]
JFYI, despite adding commits, which seem to make `update-buffer.lua` ready, it's actually not ready yet) I bumped into a case, about which I thought before, but couldn't reproduce. And now I finally can and need to fix it. I want to rethink how char diffs should actually be applied because of this.

**Comment by faergeek** [comment]
I reimplemented charwise hunk application. Now `nvim_buf_set_text` is used for all "sub" hunks. This way extmarks have more chances to survive formatting :-) There's also a chance that cursor adjustments will not be needed in the next neovim release (see TODO comment in `update_buffer_with_charwise_diff`). Probably even marks will be properly adjusted as well instead of being deleted.

I also tried to measure how bad it performs in extreme cases. And it's surprisingly good actually. The case was formatting minified jquery lib with prettier) And with this approach it takes ~0.8 seconds on my machine. Though I also noticed that in an upcoming version of neovim it will be worse and what I saw is 4-5 seconds (related to new patch from vim being applied, where more checks happen on every change in a buffer) with almost all time being spent in `nvim_buf_set_text`. But I think this is an extreme case anyway. I can share a profiling setup for that if you're interested.

**Comment by elentok** [comment]
Hi @faergeek,

Sorry for the (super) delayed response :(
I've been swamped with work...

This looks good, have you been using it? is it safe to merge?

And thanks again for this amazing work 👍

**Comment by faergeek** [comment]
@elentok Hi! No worries, it's all good.

Yes, I used it, seems to work nicely and ready to merge. Not sure if having a second experimental flag on top of another one is ok though, but that's up to you anyway.

**Comment by elentok** [comment]
@faergeek merged!

I'll play with it for about a week and I think then I'll switch the partial_update to 'diff' by default and after a while just remove it.

Thanks again, I played with the new version a bit and it works great!

**Pull request merged**